### PR TITLE
[Amoro-2065] ActivePluginManager interface and implementation

### LINKED ISSUE #2065: [Subtask]: ActivePluginManager interface and implementation
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### Description

some plugins are factory implementation and some like metrics plugins are not. The key difference is that these plugins keep their own states and need opening and closing:

```
public interface ActivePlugin extends AmoroPlugin {

  void open(Map<String, String> properties);

  void close();
}
```

For these plugins, there should be a corresponding plugin manager named as ActivePluginManager which would open plugin and load configuration(Map<String, String>) from a specified location

### Parent issue

https://github.com/NetEase/amoro/issues/2041

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by majin1102**
## Why are the changes needed?

Close #2065

## Brief change log

some plugins are factory implementation and some like metrics plugins are not. The key difference is that these plugins keep their own states and need opening and closing:

```
public interface ActivePlugin extends AmoroPlugin {

  void open(Map<String, String> properties);

  void close();
}
```
For these plugins, there should be a corresponding plugin manager named as ActivePluginManager which would open plugin and load configuration(Map<String, String>) from a specified location

## How was this patch tested?

- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: ams/api/src/main/java/com/netease/arctic/ams/api/PluginManager.java [hunk 1]
```diff
@@ -18,13 +18,16 @@
 
 package com.netease.arctic.ams.api;
 
+
+import com.google.common.collect.Lists;
+
```
Comment (baiyangtx) [inline ams/api/src/main/java/com/netease/arctic/ams/api/PluginManager.java:23]: Using iceberg shaded guava instead.

File: ams/api/src/main/java/com/netease/arctic/ams/api/metrics/MetricType.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.netease.arctic.ams.api.metrics;
```
Comment (baiyangtx) [inline ams/api/src/main/java/com/netease/arctic/ams/api/metrics/MetricType.java:1]: License is required.

File: ams/server/src/main/java/com/netease/arctic/server/exception/LoadingPluginException.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
+package com.netease.arctic.server.exception;
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/exception/LoadingPluginException.java:1]: License is required.

File: ams/server/src/main/java/com/netease/arctic/server/utils/PreconditionUtils.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.netease.arctic.server.utils;
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/utils/PreconditionUtils.java:1]: License is required.

File: ams/server/src/main/java/com/netease/arctic/server/utils/PreconditionUtils.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    if (!exists) {
+      throw new ObjectNotExistsException(objectName);
+    }
+  }
+
+  public static void checkNotExist(boolean exists, String objectName) {
+    if (exists) {
+      throw new AlreadyExistsException(objectName);
+    }
+  }
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/utils/PreconditionUtils.java:18]: Could we use guava Preconditions instead ?

**Comment by baiyangtx** [comment]
According to the Amoro project specifications, the PR needs to briefly describe the specific changes made.

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class ActivePluginManager<T extends ActivePlugin>
+    implements PluginManager<T>, Iterable<T> {
+
+  protected static final String PUGIN_IMPLEMENTION_CLASS = "impl";
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java]: Spelling mistake.

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+  @Override
+  public void install(String pluginName) {
+    PreconditionUtils.checkNotExist(installedPlugins.containsKey(pluginName),
+        "Plugin " + pluginName);
+    Map<String, String> properties = loadProperties(pluginName);
+    String pluginClass = properties.get(PUGIN_IMPLEMENTION_CLASS);
+    try {
+      Class<?> clazz = Class.forName(pluginClass);
+      T plugin = (T) clazz.newInstance();
+      installedPlugins.computeIfAbsent(pluginName, k -> {
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java:77]: The code earlier has already performed a checkNotExist, is the use of computeIfAbsent at this point for multi-threaded scenarios? If it is for multi-threading, can we consider putting the class loading and other into the same function?

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    T plugin = installedPlugins.remove(pluginName);
+    if (plugin != null) {
+      plugin.close();
+    }
+  }
+
+  @NotNull
+  @Override
+  public Iterator<T> iterator() {
+    return new Iterator<T>() {
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java:107]: Can return 'return installedPlugins.values().iterator()' dirctly

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+import com.netease.arctic.ams.api.PluginManager;
+import com.netease.arctic.server.exception.LoadingPluginException;
+import com.netease.arctic.server.utils.PreconditionUtils;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public abstract class ActivePluginManager<T extends ActivePlugin>
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java:38]: PluginManager extends Iterable already, ActivePluginManager don't need implements Iterable again

File: ams/server/src/main/java/com/netease/arctic/server/utils/PreconditionUtils.java (hunk 5 above)
Comment (majin1102) [reply ams/server/src/main/java/com/netease/arctic/server/utils/PreconditionUtils.java:18]: guava is not suitable here

File: ams/server/src/main/java/com/netease/arctic/server/exception/LoadingPluginException.java (hunk 3 above)
Comment (majin1102) [reply ams/server/src/main/java/com/netease/arctic/server/exception/LoadingPluginException.java:19]: done

File: ams/api/src/main/java/com/netease/arctic/ams/api/metrics/MetricType.java (hunk 2 above)
Comment (majin1102) [reply ams/api/src/main/java/com/netease/arctic/ams/api/metrics/MetricType.java:19]: done

File: ams/api/src/main/java/com/netease/arctic/ams/api/PluginManager.java (hunk 1 above)
Comment (majin1102) [reply ams/api/src/main/java/com/netease/arctic/ams/api/PluginManager.java]: done

File: ams/server/src/main/java/com/netease/arctic/server/manager/SpiPluginManager.java [hunk 10]
```diff
@@ -66,12 +69,27 @@ public void install(String pluginName) {
...
+        "Plugin " + pluginName);
     installedPlugins.remove(pluginName);
   }
 
+  @NotNull
   @Override
-  public List<T> list() {
-    return new ArrayList<>(installedPlugins.values());
+  public Iterator<T> iterator() {
+    return new Iterator<T>() {
```
Comment (shidayang) [inline ams/server/src/main/java/com/netease/arctic/server/manager/SpiPluginManager.java:80]: Can return 'return installedPlugins.values().iterator()' dirctly

**Review by baiyangtx — APPROVED** [review]
LGTM, thanks for your contribution

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java (hunk 6 above)
Comment (majin1102) [reply ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java]: already fixed, check newest code

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java (hunk 8 above)
Comment (majin1102) [reply ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java:106]: this is meant to return an unmodifiable iterator

File: ams/server/src/main/java/com/netease/arctic/server/manager/SpiPluginManager.java (hunk 10 above)
Comment (majin1102) [reply ams/server/src/main/java/com/netease/arctic/server/manager/SpiPluginManager.java:80]: the same as above, unmodifiable

File: ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java (hunk 7 above)
Comment (majin1102) [reply ams/server/src/main/java/com/netease/arctic/server/manager/ActivePluginManager.java]: yeah, that will be better

File: ams/server/src/main/java/com/netease/arctic/server/manager/SpiPluginManager.java (hunk 10 above)
Comment (shidayang) [reply ams/server/src/main/java/com/netease/arctic/server/manager/SpiPluginManager.java:80]: get

**Review by shidayang — APPROVED** [review]

**Comment by shidayang** [comment]
LGTM

**Pull request merged**